Data/ByteString/Internal.hs

```diff
@@ -357,19 +357,21 @@ unsafePackLiteral addr# =
 
 packUptoLenBytes :: Int -> [Word8] -> (ByteString, [Word8])
 packUptoLenBytes len xs0 =
-    unsafeCreateUptoN' len $ \p -> go p len xs0
-  where
-    go !_ !n []     = return (len-n, [])
-    go !_ !0 xs     = return (len,   xs)
-    go !p !n (x:xs) = poke p x >> go (p `plusPtr` 1) (n-1) xs
+    unsafeCreateUptoN' len $ \p0 ->
+      let p_end = plusPtr p0 len
+          go !p xs | p == p_end   = return (len, xs)
+          go !p xs | List.null xs = return (p `minusPtr` p0, xs)
```

Comment:
Why `List.null` instead of pattern matching on `[]`? 

I would rather stick to the original order of statements, unless there is a reason to swap.

Reply:
> Why List.null instead of pattern matching on []?

I wanted to reuse the `xs` from the left hand side in the right hand side to avoid constructing a new `[]` object. I don't mind doing the same with a `xs@[]` pattern if you'd prefer that.

> I would rather stick to the original order of statements, unless there is a reason to swap.

I put the clause that seems (pure speculation on my side) to do a bit less work as the first one.

Reply:
Yes, I would prefer `xs@[]`. Not sure if binding `[]` to `xs` makes any difference allocation-wise: I thought that all nullary constructors are allocated only once, globally. 

My understanding is that swapping clauses makes `go` lazy in the second argument, because there appears a codepath, which does not trigger even weak-head normal form of it. This is a difference in behaviour, which I would rather avoid in course of this PR, unless we are happy to spend more time investigating Haskell Core.

Reply:
> I thought that all nullary constructors are allocated only once, globally.

Interesting, I didn't know that, thank you!